gilbreth_perception/src/recognition_node.cpp

```diff
@@ -119,83 +124,185 @@ void loadModel(std::vector<std::vector<double> > &pick_pose, std::vector<std::st
     norm_est.compute(*model_normals);
     model_normals_list.push_back(model_normals);
   }
+  // if use ICP
+  if (icp) {
+    pcl::FPFHEstimation<pcl::PointXYZ, pcl::Normal, pcl::FPFHSignature33> fpfh_est;
+    for (int i = 0; i < model_list.size(); i++) {
+      pcl::PointCloud<pcl::FPFHSignature33>::Ptr model_features(new pcl::PointCloud<pcl::FPFHSignature33>);
+      fpfh_est.setSearchMethod(tree);
+      fpfh_est.setRadiusSearch(descr_rad);
+      fpfh_est.setInputCloud(model_list[i]);
+      fpfh_est.setInputNormals(model_normals_list[i]);
+      fpfh_est.compute(*model_features);
+      model_features_list.push_back(model_features);
+    }
+  }
+  // if use correspondence grouping
+  else {
+    pcl::SHOTEstimationOMP<PointType, NormalType, pcl::SHOT352> descr_est;
+    descr_est.setRadiusSearch(descr_rad);
+    pcl::BOARDLocalReferenceFrameEstimation<PointType, NormalType, pcl::ReferenceFrame> rf_est;
 
-  for (int i = 0; i < model_list.size(); i++) {
-    pcl::PointCloud<pcl::FPFHSignature33>::Ptr model_features(new pcl::PointCloud<pcl::FPFHSignature33>);
-    fpfh_est.setSearchMethod(tree);
-    fpfh_est.setRadiusSearch(descr_rad);
-    fpfh_est.setInputCloud(model_list[i]);
-    fpfh_est.setInputNormals(model_normals_list[i]);
-    fpfh_est.compute(*model_features);
-    model_features_list.push_back(model_features);
+    for (int i = 0; i < model_list.size(); i++) {
+      pcl::PointCloud<pcl::SHOT352>::Ptr model_descriptors(new pcl::PointCloud<pcl::SHOT352>());
+      descr_est.setInputCloud(model_list[i]);
+      descr_est.setInputNormals(model_normals_list[i]);
+      descr_est.setSearchSurface(model_list[i]);
+      descr_est.compute(*model_descriptors);
+      model_descriptor_list.push_back(model_descriptors);
+
+      pcl::PointCloud<pcl::ReferenceFrame>::Ptr model_rf(new pcl::PointCloud<pcl::ReferenceFrame>());
+      rf_est.setFindHoles(true);
+      rf_est.setRadiusSearch(0.015);
+      rf_est.setInputCloud(model_list[i]);
+      rf_est.setInputNormals(model_normals_list[i]);
+      rf_est.setSearchSurface(model_list[i]);
+      rf_est.compute(*model_rf);
+      model_rf_list.push_back(model_rf);
+    }
   }
 }
 
 void cloudCallBack(const sensor_msgs::PointCloud2ConstPtr &cloud_msg, const tf::TransformListener &listener,
```

Comment:
Rather than passing this many arguments in one function would it be best to make this function into a class method and set the arguments as class members?

Reply:
Yes, I was thinking about that last Friday, I am going to change it.